src/dbnode/storage/index/block.go

```diff
@@ -533,7 +533,7 @@ func (b *block) addQueryResults(
 	batch []doc.Document,
 ) ([]doc.Document, int, int, error) {
 	// update recently queried docs to monitor memory.
-	if err := b.queryStats.Update(len(batch)); err != nil {
+	if err := b.queryLimits.DocsLimit().Inc(len(batch)); err != nil {
```

Comment:
De-ref the .DocsLimit() when you cache the limits on the `*block` perhaps? This is a fairly tight loop too.